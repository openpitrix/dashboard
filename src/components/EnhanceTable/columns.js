import React from 'react';
import { I18n } from 'react-i18next';

import LessText from 'components/LessText';
import TdName from 'components/TdName';
import TimeShow from 'components/TimeShow';
import { getObjName } from 'utils';
import _ from 'lodash';

// inject
export default ({ users = [], urlPrefix = '' }) => [
  {
    title: 'App Name',
    key: 'name',
    width: '150px',
    render: item => (
      <TdName
        name={item.name}
        description={item.app_id}
        image={item.icon}
        linkUrl={urlPrefix + item.app_id}
      />
    )
  },
  {
    title: 'App intro',
    key: 'intro',
    width: '150px',
    render: item => (
      <LessText txt={item.abstraction || item.description} limit={40} />
    )
  },
  {
    title: 'Categories',
    key: 'category',
    width: '100px',
    render: item => (
      <I18n>
        {t => {
          const cateName = _.get(item, 'category_set', [])
            .filter(cate => cate.category_id && cate.status === 'enabled')
            .map(cate => cate.name)
            .join(', ');

          return t(cateName);
        }}
      </I18n>
    )
  },
  {
    title: 'Delivery type',
    key: 'delivery_type',
    width: '80px',
    render: item => <I18n>{t => item.app_version_types || t('unknown')}</I18n>
  },
  {
    title: 'Count deploy',
    key: 'cnt_deploy',
    width: '80px',
    // todo
    render: () => <I18n>{t => t('unknown')}</I18n>
  },
  {
    title: 'Developer',
    key: 'owner',
    width: '80px',
    render: item => getObjName(users, 'user_id', item.owner, 'username') || item.owner
  },
  {
    title: 'App service provider',
    key: 'maintainers',
    width: '80px',
    render: item => {
      let { maintainers } = item;
      if (maintainers && _.isString(maintainers)) {
        try {
          maintainers = _.map(JSON.parse(maintainers), 'name').join(', ');
        } catch (e) {}
      }

      return maintainers;
    }
  },
  {
    title: 'Publish time',
    key: 'status_time',
    width: '130px',
    render: item => <TimeShow time={item.status_time} />
  }
];
