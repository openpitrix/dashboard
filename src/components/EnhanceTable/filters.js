import { mappingStatus } from 'utils';

const defaultConditions = [
  { name: 'Draft', value: 'draft' },
  { name: mappingStatus('Active'), value: 'active' },
  { name: mappingStatus('Suspended'), value: 'suspended' },
  { name: 'Deleted', value: 'deleted' }
];

const filters = (onChange, selectValue, replace = []) => [
  {
    key: 'status',
    conditions: replace.length ? replace : defaultConditions,
    onChangeFilter: onChange,
    selectValue
  }
];

export default filters;
