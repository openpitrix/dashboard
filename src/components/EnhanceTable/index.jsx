import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import { Table } from 'components/Base';

import defaultColumns from './columns';
import defaultFilters from './filters';

/**
 *  A enhanced table component to simplify operations
 */
export default class EnhanceTable extends React.PureComponent {
  static propTypes = {
    canCancelPager: PropTypes.bool,
    className: PropTypes.string,
    columns: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
    columnsFilter: PropTypes.func,
    data: PropTypes.array,
    filterList: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
    inject: PropTypes.object,
    isLoading: PropTypes.bool,
    replaceFilterConditions: PropTypes.array,
    selectionType: PropTypes.string,
    store: PropTypes.shape({
      selectedRowKeys: PropTypes.array,
      selectIds: PropTypes.array,
      selectStatus: PropTypes.string,
      searchWord: PropTypes.string,
      totalCount: PropTypes.number,
      currentPage: PropTypes.number,
      onChangeSelect: PropTypes.func,
      onChangeStatus: PropTypes.func,
      changePagination: PropTypes.func
    }).isRequired,
    tableType: PropTypes.string
  };

  static defaultProps = {
    store: {
      selectedRowKeys: [],
      onChangeSelect: _.noop,
      onChangeStatus: _.noop,
      selectStatus: '',
      changePagination: _.noop,
      totalCount: 0,
      currentPage: 1
    },
    data: [],
    columns: [],
    columnsFilter: x => x,
    selectionType: 'checkbox',
    tableType: 'Apps',
    filterList: [],
    inject: {},
    canCancelPager: true,
    replaceFilterConditions: []
  };

  render() {
    const {
      data,
      store,
      columns,
      columnsFilter,
      inject,
      filterList,
      selectionType,
      tableType,
      isLoading,
      className,
      canCancelPager,
      replaceFilterConditions
    } = this.props;

    const rowSelection = {
      type: selectionType,
      selectedRowKeys: store.selectedRowKeys,
      onChange: store.onChangeSelect
    };

    const pagination = {
      tableType,
      onChange: store.changePagination,
      total: store.totalCount,
      current: store.currentPage,
      noCancel: !canCancelPager
    };

    let normalizeCols = _.isEmpty(columns) ? defaultColumns : columns;

    if (_.isFunction(normalizeCols)) {
      normalizeCols = normalizeCols(inject);
    }

    return (
      <Table
        className={className}
        isLoading={isLoading}
        columns={columnsFilter(normalizeCols)}
        dataSource={data.slice(0, store.pageSize || 10)}
        rowSelection={rowSelection}
        filterList={
          _.isEmpty(filterList)
            ? defaultFilters(
              store.onChangeStatus,
              store.selectStatus,
              replaceFilterConditions
            )
            : filterList
        }
        pagination={pagination}
      />
    );
  }
}
