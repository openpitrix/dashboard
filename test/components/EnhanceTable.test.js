import { BrowserRouter } from 'react-router-dom';

import EnhanceTable from 'components/EnhanceTable';

describe('Component EnhanceTable ', () => {
  it('basic render', () => {
    const wrapper = render(
      <BrowserRouter>
        <EnhanceTable />
      </BrowserRouter>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
