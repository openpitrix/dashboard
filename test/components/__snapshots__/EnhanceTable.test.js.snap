// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component EnhanceTable  basic render 1`] = `
<div
  class="table"
>
  <div
    class="pi-table pi-table-scroll-position-left"
  >
    <div
      class="pi-table-content"
    >
      <div
        class="pi-table-body"
      >
        <table
          class=""
        >
          <colgroup>
            <col
              style="width:150px;min-width:150px"
            />
            <col
              style="width:150px;min-width:150px"
            />
            <col
              style="width:100px;min-width:100px"
            />
            <col
              style="width:80px;min-width:80px"
            />
            <col
              style="width:80px;min-width:80px"
            />
            <col
              style="width:120px;min-width:120px"
            />
            <col
              style="width:120px;min-width:120px"
            />
            <col
              style="width:130px;min-width:130px"
            />
          </colgroup>
          <thead
            class="pi-table-thead"
          >
            <tr>
              <th
                class=""
              >
                App Name
              </th>
              <th
                class=""
              >
                App intro
              </th>
              <th
                class=""
              >
                Categories
              </th>
              <th
                class=""
              >
                Delivery type
              </th>
              <th
                class=""
              >
                Count deploy
              </th>
              <th
                class=""
              >
                Developer
              </th>
              <th
                class="boldFont"
              >
                App service provider
              </th>
              <th
                class=""
              >
                Publish time
              </th>
            </tr>
          </thead>
          <tbody
            class="pi-table-tbody"
          />
        </table>
      </div>
      <div
        class="pi-table-placeholder"
      >
        <div
          class="noData"
        >
          <span
            class="icon icon-appcenter"
            style="width:84px;height:84px"
          >
            <svg
              class="qicon qicon-appcenter qicon-dark"
            >
              <use
                href="#qui-appcenter"
              />
            </svg>
          </span>
          <div
            class="word"
          >
            No data!
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
